File: Baekjoon/Python/BaekJoon2493/탑.py
# /*
#  * BaekJoon 2493 탑
#  * programmer: yooj
#  * Date: 21.07.26
#  * using: Pycharm  & Python3
#  * Site: https://www.acmicpc.net/problem/2493 탑
#  */
import sys

n = int(sys.stdin.readline())  # 탑의 개수를 입력 받음
stk1 = []  # 처음 인덱스와 값을 저장하기 위한 스택
stk2 = []  # 답을 체크하기 위한 스택
ans = [0] * (n + 1)  # 0으로 초기화 함
data = list(map(int, sys.stdin.readline().split()))  # 한 줄을 입력 ㅂ다음
for i in range(n):  # 스택에 입력
    stk1.append((i + 1, data[i]))
while stk1:  # 스택 1이 비지 않았으면
    if len(stk1) == 1:  # stk1 의 길이 1이면 그 탑은 전송할 곳이 없음
        break
    stk2.append((stk1.pop()))  # 스택 2에 스택 1의 최상단을 추가
    while stk2 and stk1[-1][1] > stk2[-1][1]:  # stk2 가 존재하고 다음 탑이 이전 탑들보다 길경우 정답을 입력함
        ans[stk2[-1][0]] = stk1[-1][0]
        stk2.pop()  # 답을 입력한 타워는 제거

sys.stdout.write(' '.join(map(str, ans[1:])))  # 정답 출력

File: Baekjoon/C++/BaekJoon2493/탑.cpp
/*
* BaekJoon 2493 탑
* programmer: yooj
* Date: 21.07.26
* using: Pycharm  & Python3
* Site: https://www.acmicpc.net/problem/2493 탑
*/


#include <iostream>
#include <stack>

using namespace std;


stack<pair<int, int>> stk1;  // 기존 배열을 저장하기 위한 스택
stack<pair<int, int>> stk2;  // 비교할 대상을 저작하기 위한 스택
int main() {
    ios::sync_with_stdio(false);
    cout.tie(nullptr);
    cin.tie(nullptr);
    int n;
    cin >> n;  // 탑의 개수를 저장함
    int ans[n + 1];  // 정답을 저장하기 위한 배열
    fill(ans, ans + n, 0);//0 으로 초기화
    for (int i = 0; i < n; i++) {// 스택에 값을 입력함
        int t;
        cin >> t;
        stk1.push(make_pair(i + 1, t));
    }
    while (!stk1.empty()) {// 스택이 비지 않으면 계속함
        if (stk1.size() == 1)break;// 맨 마지막 탑은 항상 불가능함
        stk2.push(make_pair(stk1.top().first, stk1.top().second));//스택 2에 스택1의 최상단을 push
        stk1.pop();
        while (!stk2.empty() && stk1.top().second > stk2.top().second) {//비교를 해서 해당할경우 답을 추가함
            ans[stk2.top().first] = stk1.top().first;
            stk2.pop();
        }
    }
    for (int i = 1; i < n + 1; i++) {//정답출력
        cout << ans[i] << ' ';
    }
}